### PR TITLE
Add complex number support to `asarray`

### PR DESCRIPTION
**Pull request by kgryte**
This PR

-   adds complex number support to `asarray`. Importantly, if one or more values are `complex` numbers, the output data type must be the default complex floating-point data type.

### CODE REVIEW

File: spec/API_specification/array_api/creation_functions.py
Comment (rgommers) [inline spec/API_specification/array_api/creation_functions.py:46]: The precedence doesn't seem needed, there's only ever one condition that can be true.

File: spec/API_specification/array_api/creation_functions.py
Comment (rgommers) [inline spec/API_specification/array_api/creation_functions.py:49]: This one is a little odd, you can read it as explicitly disallowing `int`-only values.

**Review by rgommers — COMMENTED** [review]
LGTM modulo a few small textual things.

File: spec/API_specification/array_api/creation_functions.py
Comment (asmeurer) [reply spec/API_specification/array_api/creation_functions.py:46]: I think that was the case before but now the last two bullet points can both be true.

File: spec/API_specification/array_api/creation_functions.py
Comment (kgryte) [reply spec/API_specification/array_api/creation_functions.py:49]: Will rephrase.

Comment (kgryte) [reply spec/API_specification/array_api/creation_functions.py]: This has now been addressed.

**Comment by leofang** [comment]
I fail to see why complex should appear before (or after) float with the "precedence" clause, I must be missing something...

**Comment by kgryte** [comment]
@leofang Because if at least one value is `complex`, then the returned array must have a complex data type. Hence, if provided a mixture of `complex` and `float`, `complex` takes precedence. This follows type promotion semantics.

**Review by leofang — APPROVED** [review]

**Comment by kgryte** [comment]
Thanks, @leofang! As this now has two approvals, will merge and any further changes can be addressed through subsequent PRs...

**Pull request merged**
